File: books/algorithms/rust/quickSort.rs
/** states quickSort in a clear way
 * 1. set a guard, such as the first element
 * 2. for every other element, if smaller, move to left, larger 
 * move to right
 * 3. swap the guard and the last index move.
 * 
 */

fn partition(mut a: &[u32; 12], p: u32, r: u32) {
    // Not the p is not the begin and r is not the length of A!
    // set the first element as a guard
    let x = a[p];

}


fn main() {
    println!("QuickSort in Rust!");
    let mut a = [2, 8, 12, 7, 9, 14, 5, 6, 16, 1, 3, 10];
    partition(&mut a, 0, a.len().count_ones());
}

File: books/algorithms/cpp/quicksort.cpp
// quicksort in C++
#include <vector>
#include <iostream>

using namespace std;


void swap(vector<int>& vec, int i, int j) 
{
    int t = vec.at(i);
    vec[i] = vec.at(j);
    vec[j] = t;
}

int partition(vector<int>& vec, int start, int end) 
{
    int x = vec.at(end-1);
    int i = start;

    for (int j = start; j < end - 1; j++) {
        if (vec.at(j) < x) {
            swap(vec, j, i);
            i++;
        }
    }
    swap(vec, i, end-1);
    return i;
}

void quicksort(vector<int>& vec, int start, int end) 
{
    if (start < end) {
        int q = partition(vec, start, end);
        quicksort(vec, start, q);
        quicksort(vec, q+1, end);
    }
}

void print(vector<int>& vec)
{
    for (int i : vec) {
        cout << i << " ";
    }
    cout << endl;
}


int main()
{
    vector<int> vec{2, 8, 12, 7, 9, 14, 5, 6, 16, 1, 3, 10};
    quicksort(vec, 0, vec.size());
    print(vec);
}